x-pack/plugins/ml/server/models/job_service/datafeeds.ts

```diff
@@ -204,12 +208,160 @@ export function datafeedsProvider(mlClient: MlClient) {
     }
   }
 
+  async function datafeedPreview(job: Job, datafeed: Datafeed) {
+    let query: any = { match_all: {} };
+    if (datafeed.query) {
+      query = datafeed.query;
+    }
+    const { getTimeFieldRange } = fieldsServiceProvider(client);
+    const { start } = await getTimeFieldRange(
+      datafeed.indices,
+      job.data_description.time_field,
+      query,
+      datafeed.indices_options
+    );
+
+    // Get bucket span
+    // Get first doc time for datafeed
+    // Create a new query - must user query and must range query.
+    // Time range 'to' first doc time plus < 10 buckets
+
+    // Do a preliminary search to get the date of the earliest doc matching the
+    // query in the datafeed. This will be used to apply a time range criteria
+    // on the datafeed search preview.
+    // This time filter is required for datafeed searches using aggregations to ensure
+    // the search does not create too many buckets (default 10000 max_bucket limit),
+    // but apply it to searches without aggregations too for consistency.
+    const bucketSpan = parseInterval(job.analysis_config.bucket_span);
+    if (bucketSpan === null) {
+      return;
+    }
+    const earliestMs = start.epoch;
+    const latestMs = +start.epoch + 10 * bucketSpan.asMilliseconds();
+
+    const body: any = {
+      query: {
+        bool: {
+          must: [
+            {
+              range: {
+                [job.data_description.time_field]: {
+                  gte: earliestMs,
+                  lt: latestMs,
+                  format: 'epoch_millis',
+                },
+              },
+            },
+            query,
+          ],
+        },
+      },
+    };
+
+    // if aggs or aggregations is set, add it to the search
+    const aggregations = datafeed.aggs ?? datafeed.aggregations;
+    if (aggregations && Object.keys(aggregations).length) {
+      body.size = 0;
+      body.aggregations = aggregations;
+
+      // add script_fields if present
+      const scriptFields = datafeed.script_fields;
+      if (scriptFields && Object.keys(scriptFields).length) {
```

Comment:
We can probably use `isPopulatedObject` for these checks.

Reply:
https://github.com/elastic/kibana/pull/91830/commits/8ca21bb9de9c182d0c95e40cd550f4596af8f8e0